src/greenlight/runner.clj

```diff
@@ -77,6 +102,68 @@
     true))
 
 
+(defmacro ^:private with-delayed-output
+  [printer & body]
+  `(let [out# (java.io.StringWriter.)
+         original-out# *out*
+         original-err# *err*
+         printer# ~printer]
+     (binding [*out* out#, *err* out#]
+       (let [result# (do ~@body)]
+         (binding [*out* original-out#
+                   *err* original-err#]
+           (printer# (str out#))
+           result#)))))
+
+
+(defn- sync-printer
+  "Creates a synchronized printer function"
+  []
+  (let [o (Object.)]
+    (fn [s]
+      (locking o
+        (print s)
+        (flush)))))
+
+
+(defn- execute-parallel
+  "Run a collection of tests, using an executor pool with `n-threads`"
+  [system tests n-threads]
+  (let [exec-pool (Executors/newFixedThreadPool n-threads)
+        printer (sync-printer)
+        bindings (get-thread-bindings)
+        run-group (fn run-group
+                    [tests]
+                    (reify Callable
+                      (call [_]
```

Comment:
I suspect you might just need the type-hint instead of this whole reification. The `submit` function takes both `Runbable` and `Callable<T>` in the single-argument signatures. Clojure functions implement both, for the zero-arity and one-arity signatures of `invoke`.